pkg/admission/customresourcedefinition/admission_test.go

```diff
@@ -0,0 +1,194 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customresourcedefinition
+
+import (
+	"context"
+	"testing"
+
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+
+	"github.com/kcp-dev/kcp/pkg/admission/helpers"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/reconciler/apis/apibinding"
+)
+
+func createAttr(obj *apiextensions.CustomResourceDefinition) admission.Attributes {
+	return admission.NewAttributesRecord(
+		obj,
+		nil,
+		apiextensionsv1.Kind("CustomResourceDefinition").WithVersion("v1"),
+		"",
+		"test",
+		apiextensionsv1.Resource("customresourcedefinitions").WithVersion("v1"),
+		"",
+		admission.Create,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func createAttrApiBinding(apiBinding *apisv1alpha1.APIBinding) admission.Attributes {
+	return admission.NewAttributesRecord(
+		helpers.ToUnstructuredOrDie(apiBinding),
+		nil,
+		apisv1alpha1.Kind("APIBinding").WithVersion("v1alpha1"),
+		"",
+		apiBinding.Name,
+		apisv1alpha1.Resource("apibindings").WithVersion("v1alpha1"),
+		"",
+		admission.Create,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func updateAttr(obj, old *apiextensions.CustomResourceDefinition) admission.Attributes {
+	return admission.NewAttributesRecord(
+		obj,
+		old,
+		apiextensionsv1.Kind("CustomResourceDefinition").WithVersion("v1"),
+		"",
+		"test",
+		apiextensionsv1.Resource("customresourcedefinitions").WithVersion("v1"),
+		"",
+		admission.Update,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func TestValidate(t *testing.T) {
+	tests := []struct {
+		name        string
+		attr        admission.Attributes
+		clusterName string
+
+		wantErr bool
+	}{
+		{
+			name: "passes create no annotation",
+			attr: createAttr(&apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+			}),
+		},
+		{
+			name: "fails create annotation wrong workspace",
+			attr: createAttr(&apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+					Annotations: map[string]string{
+						boundCRDWorkspace: "true",
+					},
+				},
+			}),
+			wantErr: true,
+		},
+		{
+			name: "passes create annotation correct workspace",
+			attr: createAttr(&apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+					Annotations: map[string]string{
+						"apis.kcp.dev/bound-crd": "true",
```

Comment:
Use the constant for the key

Reply:
actually using the literal makes more sense here, or in tests in general. That causes pain when these change and that's good.